src/soar_sdk/views/view_parser.py

```diff
@@ -0,0 +1,123 @@
+from typing import (
+    Any,
+    Callable,
+    TypeVar,
+    Union,
+    get_origin,
+    get_args,
+    Generic,
+    cast,
+)
+import contextlib
+import inspect
+from soar_sdk.action_results import ActionOutput, ActionResult
+from soar_sdk.models.view import ViewContext, AppRunMetadata
+
+T = TypeVar("T", bound=ActionOutput)
+
+
+AllAppRuns = list[tuple[AppRunMetadata, list[ActionResult]]]
+
+
+class ViewFunctionParser(Generic[T]):
+    """Handles parsing and validation of view function signatures and execution."""
+
+    def __init__(
+        self,
+        function: Callable,
+    ) -> None:
+        self.function = function
+
+        # Auto-detect output class from function signature
+        detected_class = self.auto_detect_output_class(function)
+        self.output_class: type[T] = cast(type[T], detected_class)
+
+    @staticmethod
+    def auto_detect_output_class(function: Callable) -> type[ActionOutput]:
+        """Auto-detect ActionOutput class from function type annotations."""
+        signature = inspect.signature(function)
+
+        for param in signature.parameters.values():
+            if param.annotation == inspect.Parameter.empty:
+                continue
+
+            origin = get_origin(param.annotation)
+            if origin is list:
+                args = get_args(param.annotation)
+                if args and issubclass(args[0], ActionOutput):
+                    return args[0]
+            elif isinstance(param.annotation, type) and issubclass(
+                param.annotation, ActionOutput
+            ):
+                return param.annotation
+
+        raise TypeError(
+            f"Could not auto-detect ActionOutput class from function signature of {function.__name__}."
+        )
+
+    def parse_action_results(
+        self, raw_all_app_runs: list[tuple[Any, list[ActionResult]]]
+    ) -> tuple[list[T], list[tuple[AppRunMetadata, list[ActionResult]]]]:
+        parsed_outputs: list[T] = []
+        parsed_app_runs: list[tuple[AppRunMetadata, list[ActionResult]]] = []
+
+        for app_run_metadata, action_results in raw_all_app_runs:
+            if isinstance(app_run_metadata, dict):
+                if app_run_metadata:
+                    parsed_metadata: AppRunMetadata = AppRunMetadata.parse_obj(
+                        app_run_metadata
+                    )
+                else:
+                    parsed_metadata = AppRunMetadata()
```

Comment:
If `AppRunMetadata.parse_obj({})` not the same as `AppRunMetadata()`? Seems to me like these could be collapsed:
```suggestion
                parsed_metadata: AppRunMetadata = AppRunMetadata.parse_obj(
                    app_run_metadata
                )
```

Reply:
Yeah this area was improved